File: src/vibe.js
// vibe.js — the per-channel CRT/VHS look (L-V1).
//
// V1 is the NO-SHADER tier, and it ships everywhere today: 4:3 cropping,
// scanlines, vignette, grain and dead pixels are all overlay work, so the
// browser TV and the Pi get them with no change to the video pipeline. The
// shader tier (V2) needs a Metal rework on Apple and is parked in P5 — nothing
// here anticipates it beyond leaving the door open.
//
// ── the one structural decision ──────────────────────────────────────────────
// A vibe is stored as a SELF-CONTAINED JSON DOCUMENT, not as a spread of
// columns. That is deliberate: the open question is whether Vibe eventually
// supports PER-ITEM overrides (a tape in Track M carrying its own worn look),
// and the phase notes are explicit that retrofitting that later is much worse
// than allowing for it now.
//
// With a document, per-item support is "put the same shape somewhere else and
// add one step to resolve()" — a media column, or a field in a pack manifest.
// With columns it would be a schema redesign every time a new scope appears.
// So the MODEL is scope-agnostic from day one while the UI exposes only the
// global default and per-channel, which is all anyone has asked for.
//
// Resolution order (first non-null wins, field by field):
//     item (not yet exposed) → channel → global default → OFF

/** Every knob, with the value that means "do nothing". */
export const VIBE_OFF = {
  crop43: false,      // pillarbox to 4:3 — the biggest single "it looks right" win
  scanlines: 0,       // 0–1, opacity of the line pattern
  vignette: 0,        // 0–1, corner darkening
  grain: 0,           // 0–1, how MUCH static — density, not size
  grainSize: 1,       // 1–4, how COARSE that static is (1 = fine film grain, 4 = chunky VHS)
  deadPixels: 0,      // count of fixed stuck pixels, 0–12
  bleed: 0,           // 0–1, chroma bleed (saturation lift)
  bars: 0,            // 0–1, rolling hum bar — the band that drifts up a badly-earthed set
  chromaShift: 0,     // 0–1, colour fringing (RGB offset). NOT RENDERED ON APPLE — see below.
};

// ── knobs the platforms disagree about ──────────────────────────────────────
// Everything above is an overlay EXCEPT the two that need the decoded pixels:
// `bleed` and `chromaShift`. The browser can do both with CSS filters on the
// video element; Apple cannot, because VLCKit owns its output surface, and
// touching it is the V2/Metal tier (P5).
//
// This is exported rather than hardcoded in the UI so the control panel can say
// "not on Apple TV yet" next to the exact knobs that mean it, instead of
// offering a slider that silently does nothing on the device you are watching.
export const VIBE_PIXEL_KNOBS = ['bleed', 'chromaShift'];

/** Presets, so nobody has to dial six numbers to get somewhere good. */
export const VIBE_PRESETS = {
  off: { ...VIBE_OFF },
  // A tidy set in good condition — the look most people mean by "CRT".
  crt: { crop43: true, scanlines: 0.22, vignette: 0.35, grain: 0.05, grainSize: 1,
         deadPixels: 0, bleed: 0.15, bars: 0, chromaShift: 0.1 },
  // A tape that has been through the machine a few hundred times. Coarser grain
  // and a slow hum bar are most of what separates "tape" from "tube".
  vhs: { crop43: true, scanlines: 0.14, vignette: 0.3, grain: 0.16, grainSize: 2.5,
         deadPixels: 2, bleed: 0.35, bars: 0.25, chromaShift: 0.35 },
  // A set in the corner of a bar with a bent aerial.
  rough: { crop43: true, scanlines: 0.3, vignette: 0.5, grain: 0.3, grainSize: 3.5,
           deadPixels: 6, bleed: 0.45, bars: 0.5, chromaShift: 0.5 },
};

const clamp01 = (v) => Math.max(0, Math.min(1, Number(v) || 0));

/** Coerce anything into a valid vibe, dropping unknown keys. */
export function normalizeVibe(input) {
  if (!input || typeof input !== 'object') return null;
  return {
    crop43: !!input.crop43,
    scanlines: clamp01(input.scanlines),
    vignette: clamp01(input.vignette),
    grain: clamp01(input.grain),
    // Size is a MULTIPLIER, not a 0–1 intensity, so it gets its own clamp.
    // Defaults to 1 rather than 0: a grain size of zero is not a look, it is an
    // invisible bug, and an older document that predates this knob must come
    // back as fine grain rather than none.
    grainSize: Math.max(1, Math.min(4, Number(input.grainSize) || 1)),
    deadPixels: Math.max(0, Math.min(12, Math.round(Number(input.deadPixels) || 0))),
    bleed: clamp01(input.bleed),
    bars: clamp01(input.bars),
    chromaShift: clamp01(input.chromaShift),
  };
}

/**
 * Merge the scopes, most specific first. Each is a full document or null.
 * Kept as a plain reduce so adding an `item` scope later is one more argument.
 */
export function resolveVibe(...scopes) {
  const out = { ...VIBE_OFF };
  for (const s of scopes.filter(Boolean).reverse()) Object.assign(out, normalizeVibe(s));
  return out;
}

/** Is this vibe doing anything at all? Lets the player skip the whole layer. */
export const vibeIsActive = (v) =>
  !!v && (v.crop43 || v.scanlines > 0 || v.vignette > 0 || v.grain > 0
          || v.deadPixels > 0 || v.bleed > 0 || v.bars > 0 || v.chromaShift > 0);

export function parseVibe(json) {
  if (!json) return null;
  try { return normalizeVibe(typeof json === 'string' ? JSON.parse(json) : json); }
  catch { return null; }
}

File: apple/dumbTVCore/Sources/dumbTVCore/Vibe.swift
import Foundation

/// The per-channel CRT/VHS look (L-V1) — the Swift half of `src/vibe.js`.
///
/// ── why this file exists ────────────────────────────────────────────────────
/// Vibe shipped on the browser TV only. The model, the presets, the storage and
/// the live application were all there; the Apple apps had no vibe code at all.
/// So an owner who set a look in the web UI and then watched an Apple TV saw
/// nothing happen and reasonably concluded the feature was broken. It wasn't —
/// it was simply absent on the platform they were looking at.
///
/// ── the structural decision, inherited deliberately ─────────────────────────
/// A vibe is a SELF-CONTAINED DOCUMENT, not a spread of columns, and resolution
/// is a LIST of scopes rather than a fixed pair. That is what makes per-ITEM
/// overrides (a Track M tape carrying its own worn look) one more argument to
/// `resolve` instead of a schema change. Owner confirmed 2026-08-03 that
/// per-item is allowed in the model, so this shape is now load-bearing.
///
///     resolution: item (not yet exposed) → channel → global default → off
///
/// ── what Apple can and cannot do ────────────────────────────────────────────
/// Everything here is COMPOSITING — layers drawn over the picture — so it works
/// with VLCKit's video output untouched, exactly like the CSS overlays do in the
/// browser. `bleed` is the exception: nudging chroma means touching the decoded
/// pixels, and VLCKit owns those. It stays a no-op on Apple until the V2 Metal
/// tier (P5), and the web UI labels it accordingly rather than offering a knob
/// that silently does nothing.
public struct Vibe: Codable, Hashable, Sendable {
    /// Pillarbox to 4:3 — the phase notes call this the single biggest
    /// "it looks right" win, and it is one rule.
    public var crop43: Bool
    /// 0–1, opacity of the line pattern.
    public var scanlines: Double
    /// 0–1, corner darkening.
    public var vignette: Double
    /// 0–1, persistent static over the picture.
    public var grain: Double
    /// Count of fixed stuck pixels, 0–12.
    public var deadPixels: Int
    /// 0–1, chroma bleed. **Not rendered on Apple** — see the note above.
    public var bleed: Double

    public init(crop43: Bool = false, scanlines: Double = 0, vignette: Double = 0,
                grain: Double = 0, deadPixels: Int = 0, bleed: Double = 0) {
        self.crop43 = crop43
        self.scanlines = Self.clamp01(scanlines)
        self.vignette = Self.clamp01(vignette)
        self.grain = Self.clamp01(grain)
        self.deadPixels = max(0, min(12, deadPixels))
        self.bleed = Self.clamp01(bleed)
    }

    /// Every knob at the value that means "do nothing".
    public static let off = Vibe()

    /// Presets, so nobody has to dial six numbers to get somewhere good.
    /// Values are copied from `src/vibe.js` and must stay in lockstep — the web
    /// UI names a preset and both engines have to agree what that name means.
    public static let presets: [String: Vibe] = [
        "off": .off,
        // A tidy set in good condition — the look most people mean by "CRT".
        "crt": Vibe(crop43: true, scanlines: 0.22, vignette: 0.35, grain: 0.05, deadPixels: 0, bleed: 0.15),
        // A tape that has been through the machine a few hundred times.
        "vhs": Vibe(crop43: true, scanlines: 0.14, vignette: 0.30, grain: 0.16, deadPixels: 2, bleed: 0.35),
        // A set in the corner of a bar with a bent aerial.
        "rough": Vibe(crop43: true, scanlines: 0.30, vignette: 0.50, grain: 0.30, deadPixels: 6, bleed: 0.45),
    ]

    /// Does this vibe do anything at all? Used to skip the whole overlay stack,
    /// so an unstyled channel costs nothing per frame.
    public var isActive: Bool {
        crop43 || scanlines > 0 || vignette > 0 || grain > 0 || deadPixels > 0
        // `bleed` deliberately excluded: it renders nothing here, so a vibe that
        // sets only bleed must not switch the overlay on to draw an empty layer.
    }

    /// First non-nil scope wins, most-specific first — the same shape as
    /// `resolveVibe(...scopes)` in tv.js.
    ///
    /// That function merges each scope onto `off` in reverse order, which is
    /// EQUIVALENT to first-wins here because both sides only ever store complete
    /// documents (`normalizeVibe` fills every key; the decoder below defaults
    /// every key). Worth knowing they are equivalent by that invariant and not
    /// by construction — if partial documents ever became a thing, this would
    /// need to merge rather than pick.
    public static func resolve(_ scopes: Vibe?...) -> Vibe {
        for s in scopes { if let s { return s } }
        return .off
    }

    /// Decode a stored document. Anything unparseable resolves to "no vibe set"
    /// rather than throwing — a corrupt setting must never stop the television.
    public static func parse(_ json: String?) -> Vibe? {
        guard let json, !json.isEmpty, let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(Vibe.self, from: data)
    }

    /// Encode for storage. Matches the shape `normalizeVibe()` writes in Node.
    public func encoded() -> String? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// Serialise for a JSON response — the web UI reads these to fill its
    /// controls, and its preset dropdown compares against `vibePresets`.
    public var asDictionary: [String: Any] {
        ["crop43": crop43, "scanlines": scanlines, "vignette": vignette,
         "grain": grain, "deadPixels": deadPixels, "bleed": bleed]
    }

    /// Coerce a decoded-JSON value from a request body, the way `normalizeVibe`
    /// does on the Node side: a dictionary becomes a clamped vibe, and anything
    /// else — including an explicit `null` — becomes nil, which callers store as
    /// "no vibe set" so the next scope up applies.
    public static func fromAny(_ any: Any?) -> Vibe? {
        guard let d = any as? [String: Any] else { return nil }
        func num(_ k: String) -> Double {
            if let v = d[k] as? Double { return v }
            if let v = d[k] as? Int { return Double(v) }
            if let v = d[k] as? NSNumber { return v.doubleValue }
            return 0
        }
        return Vibe(crop43: (d["crop43"] as? Bool) ?? ((d["crop43"] as? NSNumber)?.boolValue ?? false),
                    scanlines: num("scanlines"), vignette: num("vignette"), grain: num("grain"),
                    deadPixels: Int(num("deadPixels").rounded()), bleed: num("bleed"))
    }

    /// Fixed positions for stuck pixels, as percentages of the picture.
    ///
    /// The SAME sequence as tv.js, because a stuck pixel is a property of the
    /// set: switching from the browser TV to the Apple TV should not move it.
    /// And they never move between launches either — a stuck pixel that wanders
    /// is not a stuck pixel.
    public static let deadPixelSpots: [(x: Double, y: Double)] = [
        (17, 23), (62, 11), (38, 77), (84, 44), (9, 58), (71, 89),
        (46, 31), (93, 67), (28, 6), (55, 95), (77, 19), (12, 41),
    ]

    private static func clamp01(_ v: Double) -> Double {
        v.isFinite ? max(0, min(1, v)) : 0
    }

    // MARK: - tolerant decoding
    //
    // Every key is optional on the way IN, defaulting to off, and the values run
    // back through `init` so they are clamped. Synthesised Codable would reject a
    // document missing a key and `parse` would then discard the whole vibe — so
    // one knob added on the Node side, or one hand-edited setting, would silently
    // turn a channel's look off entirely instead of ignoring the part it doesn't
    // understand. Unknown keys are ignored for the same reason: the web UI is
    // expected to grow knobs (grain size, hum bars, chroma shift) and an older
    // build must keep rendering the ones it does know.
    private enum CodingKeys: String, CodingKey {
        case crop43, scanlines, vignette, grain, deadPixels, bleed
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            crop43: (try? c.decodeIfPresent(Bool.self, forKey: .crop43)) .flatMap { $0 } ?? false,
            scanlines: (try? c.decodeIfPresent(Double.self, forKey: .scanlines)).flatMap { $0 } ?? 0,
            vignette: (try? c.decodeIfPresent(Double.self, forKey: .vignette)).flatMap { $0 } ?? 0,
            grain: (try? c.decodeIfPresent(Double.self, forKey: .grain)).flatMap { $0 } ?? 0,
            deadPixels: (try? c.decodeIfPresent(Int.self, forKey: .deadPixels)).flatMap { $0 } ?? 0,
            bleed: (try? c.decodeIfPresent(Double.self, forKey: .bleed)).flatMap { $0 } ?? 0)
    }
}
